File: Python/6py 3/stat_assoc.py
"""
|--------------------------------------------------------------------------
| Statistics for an Athletic Association
|--------------------------------------------------------------------------
|
"""

def to_hms(sec):
   h = str(sec // 3600).zfill(2)
   sec %= 3600
   m = str(sec // 60).zfill(2)
   s = str(sec % 60).zfill(2)
   return f"{h}|{m}|{s}"
def stat(strg):
   if strg == "": return ""
   times = []
   for t in strg.split(", "):
      h, m, s = map(int, t.split("|"))
      times.append(h * 3600 + m * 60 + s)
   times.sort()
   range_ = times[-1] - times[0]
   avg = sum(times) // len(times)
   med = (times[len(times) // 2 - 1] + times[len(times) // 2]) // 2 if len(times) % 2 == 0 else times[len(times) // 2]
   return f"Range: {to_hms(range_)} Average: {to_hms(avg)} Median: {to_hms(med)}"


from statistics import mean, median
def stat(strg):
   if not strg: return ''
   s = [sum(x * y for x,y in zip([3600, 60, 1], map(int, i.split('|')))) for i in strg.split(', ')]
   f = lambda x: f'{int(x//3600):02}|{int(x%3600//60):02}|{int(x%60):02}'
   return f'Range: {f(max(s) - min(s))} Average: {f(mean(s))} Median: {f(median(s))}'


print(
   stat("01|15|59, 1|47|16, 01|17|20, 1|32|34, 2|17|17"),
   "Range: 01|01|18 Average: 01|38|05 Median: 01|32|34"
)

File: JavaScript/6js 2/statAssoc.js
/*
|--------------------------------------------------------------------------
| Statistics for an Athletic Association
|--------------------------------------------------------------------------
|
*/

function toHMS(sec) {
   let h = String(Math.floor(sec / 3600)).padStart(2, "0");
   sec %= 3600;
   let m = String(Math.floor(sec / 60)).padStart(2, "0");
   let s = String(sec % 60).padStart(2, "0");
   return `${h}|${m}|${s}`;
}
function stat(strg) {
   if (strg === "") return "";
   let times = strg.split(", ").map((t) => {
      let [h, m, s] = t.split("|").map(Number);
      return h * 3600 + m * 60 + s;
   });
   times.sort((a, b) => a - b);
   let range = times[times.length - 1] - times[0];
   let avg = Math.floor(times.reduce((a, b) => a + b) / times.length);
   let med =
      times.length % 2 === 0
         ? Math.floor((times[times.length / 2 - 1] + times[times.length / 2]) / 2)
         : times[Math.floor(times.length / 2)];
   return `Range: ${toHMS(range)} Average: ${toHMS(avg)} Median: ${toHMS(med)}`;
}


function stat(a) {
   const toFormat = (a) =>
      [(a / 3600) | 0, (a / 60) % 60 | 0, a % 60 | 0].map((a) => ("0" + a).slice(-2)).join`|`;
   const reduce = (a) => a.reduce((a, b) => a + b, 0) / a.length;
   const arr = a.split`, `
      .map((a) => a.split`|`.reduce((a, b) => +b + a * 60, 0))
      .sort((a, b) => a - b);
   const m = arr.slice(Math.ceil(arr.length / 2) - 1, Math.floor(arr.length / 2) + 1);
   const [range, average, median] = [
      Math.max(...arr) - Math.min(...arr),
      reduce(arr),
      reduce(m),
   ].map((a) => toFormat(a));
   return a ? `Range: ${range} Average: ${average} Median: ${median}` : "";
}


console.log(
   stat("01|15|59, 1|47|16, 01|17|20, 1|32|34, 2|17|17"),
   "Range: 01|01|18 Average: 01|38|05 Median: 01|32|34"
);
console.log(
   stat("02|15|59, 2|47|16, 02|17|20, 2|32|34, 2|17|17, 2|22|00, 2|31|41"),
   "Range: 00|31|17 Average: 02|26|18 Median: 02|22|00"
);